### PR TITLE
A prototype implementation of GRPC protocol in ClickHouse.

### PR DESCRIPTION
**Pull request by mnkonkova**

### CODE REVIEW

**Comment by alexey-milovidov** [comment]
This is not meant to be merged, please switch to "draft" pull request.

**Comment by mnkonkova** [comment]
The problem may have been that the gRPC_ZLIB_PROVIDER  was not set

File: programs/server/GRPCHandler.h [hunk 1]
```diff
@@ -0,0 +1,148 @@
...
+ class CommonCallData
+ {
+     public:
+     GRPC::AsyncService* Service;
+     grpc::ServerCompletionQueue* CompilationQueue;
+     grpc::ServerContext gRPCcontext;
+     
+     IServer* iServer;
+     bool with_stacktrace = false;
+     enum CallStatus { CREATE, PREPARE, PROCESS, ONFINISH, FINISH };
```
Comment (vitlibar) [inline programs/server/GRPCHandler.h:37]: These statuses make the code complicated and hard to understand.

File: programs/server/GRPCHandler.h [hunk 2]
```diff
@@ -0,0 +1,148 @@
...
+          static_cast<CommonCallData*>(tag)->Run();
+          }
+     }
+     
+     private:
+          IServer & iServer;
+          Poco::Logger * log;
+          std::unique_ptr<grpc::ServerCompletionQueue> CompilationQueue;
+          GRPC::AsyncService Service;
+          std::unique_ptr<grpc::Server> Server;
```
Comment (vitlibar) [inline programs/server/GRPCHandler.h:143]: std::unique_ptr\<grpc::Server\> **s**erver;
See https://clickhouse.tech/docs/en/development/style/#formatting

File: programs/server/GRPCHandler.h (hunk 1 above)
Comment (vitlibar) [reply programs/server/GRPCHandler.h:37]: I looked at the function [RequestQuery](https://github.com/ClickHouse/ClickHouse/blob/2be1010cc0ab1db55674c19e6ec741d52ac41903/programs/server/GRPCHandler.cpp#L79). This function is declared in `GrpcConnection.grpc.pb.h` as follows
```
    void RequestQuery(::grpc::ServerContext* context, ::GRPCConnection::QueryRequest* request, ::grpc::ServerAsyncWriter< ::GRPCConnection::QueryResponse>* writer, ::grpc::CompletionQueue* new_call_cq, ::grpc::ServerCompletionQueue* notification_cq, void *tag) {
```
`new_call_cq` is responsible for finishing calls, and `notification_cq` is responsible for responding. You're passing the same completion queue for `new_call_cq` and for `notification_cq`. But you don't have to do so! Look:

```
GRPCServer(std::string server_address_, IServer& server_)
{
    ...
    notification_cq_ = builder.AddCompletionQueue();
    new_call_cq_.reset(new CompletionQueue());
    server = builder.BuildAndStart();
}

void HandleRpcs()
{
    new CallDataQuery(&service, new_call_cq.get(), notification_cq.get(), &iServer, log);
    
    auto handle_calls_completion = [&]()
    {
        while (true)
        {
            GPR_ASSERT(new_call_cq->Next(&tag, &ok));
            GPR_ASSERT(ok);
            delete static_cast<CallDataQuery*>(tag);
        }
    }

    auto handle_requests_completion = [&]
    {
        while (true)
        {
            GPR_ASSERT(notification_cq->Next(&tag, &ok));
            GPR_ASSERT(ok);
            auto thread = ThreadFromGlobalPool{&CallDataQuery::respond, static_cast<CallDataQuery*>(tag)};
            thread.detach();
        }

    notification_cq_thread = ThreadFromGlobalPool{handle_requests_completion};
    new_call_cq_thread = ThreadFromGlobalPool{handle_calls_completion};
}

CallDataQuery::CallDataQuery(GRPC::AsyncService* Service_, grpc::CompletionQueue* new_call_cq_, ServerCompletionQueue* notification_cq_, IServer* server_, Poco::Logger * log_)
{
Service->RequestQuery(&gRPCcontext, &request, &responder, new_call_cq, notification_cq, this);
}

CallDataQuery::respond()
{
    new CallDataQuery(Service, new_call_cq, notification_cq, iServer, log);

    context.setUser(request.user(), ...)
    ...
    parseQuery(...)
    ...
    executeQuery(...)
    ...
    responder.Write(...)
    ...
    responder.Finish()
}
```
No statuses at all!

File: programs/grpc-client/grpc_client.cpp [hunk 3]
```diff
@@ -0,0 +1,94 @@
...
+#include <memory>
+#include <string>
+
+#include <grpc++/channel.h>
+#include <grpc++/client_context.h>
+#include <grpc++/create_channel.h>
+#include <grpc++/security/credentials.h>
+#include "GrpcConnection.grpc.pb.h"
+
+class GRPCClient {
```
Comment (vitlibar) [inline programs/grpc-client/grpc_client.cpp:11]: As I already said, please convert this file into an integration unit-test.
It seems it must not be difficult, see https://grpc.io/docs/quickstart/python/

File: programs/server/GRPCHandler.h (hunk 1 above)
Comment (vitlibar) [reply programs/server/GRPCHandler.h:37]: Or, if you're making the real streaming, then call `responder.Write()` from `WriteBufferFromGRPC::nextImpl()` and call `responder.Finish()` from `WriteBufferFromGRPC::finalize()`.

**Review by vitlibar — CHANGES_REQUESTED** [review]

**Comment by vitlibar** [comment]
I'll finish the work in https://github.com/ClickHouse/ClickHouse/pull/15111

**Pull request closed without merging**